### ISSUE TITLE
TestContext should have test-output properties for Information and Warning

### ISSUE THREAD

**Issue by OsirisTerje** [nunit/nunit#2908]
TestContext currently have two properties for test-output, Error and Progress.  
1. Both write to the StdError, but only Error is a real error, Progress is informational and should be written to StdOut
2. There should be two more properties Information and Warning, allowing the user to easily add context information in the test that appears in the test output. 

Note:  I am fully aware that parallel execution makes a mess out of such messages, but this should be up to the user.  If the user have a need for these messages they should be able to use them, for diagnostic purposes or for any reason they like.  

I am adding a corresponding issue to the NUnit3TestAdapter to respond correctly to these 4 streams.  For the adapter it doesn't matter which stream is being used (StdError or StdOut) as it only see the named "streams" and the value being sent.

**Comment by CharliePoole** [comment]
Actually there are three. The third is Out, which is also the default. 

Out goes to stdout, which goes to the result. The other two are for immediate display. As you say, one is error, the other non-error.

It's incorrect to assume that both go to standard error. They go wherever the runner in use sends them. The console uses strErr but some other runner may do what it likes.

The design allows for adding more output channels but I think each one needs to be justified since every runner, nunit or third party would need to be modified to handle them.

**Comment by OsirisTerje** [comment]
Thanks Charlie! 
I find it quite natural to have Information and Warning along with Error.   I see your point about all runners need to adapt to these, but since this is an enhancements, it could be supported in upcoming releases of the runners that want to implement them?

There is of course the option of adding these as extension methods in the VS Adapter which is my current concern, but I find it cleaner to have these in NUnit itself.

**Comment by CharliePoole** [comment]
Usually, those four levels go along with conditional output, which nunit doesn't support. Console output OTOH normally only has two categories.

NUnit once had special handling for log output like log4net but we decided it was out of scope and removed it.

As an author of a couple of third party runners, I know that I'll be forced to support this feature for all practical purposes. Since I don't see anything else useful to do with it, I'll simply end up mixing the three non-error outputs back together. Seems like a lot of work to get back to where you started.

As we often tell users, could you explain the use case for this?

**Comment by nvborisenko** [comment]
Do we have a possibility to send a message from test domain to nunit addins in "silent" mode? Currently `TestContext.Progress.WriteLine()` does this trick, but messages are going into console as well. My use case is to write addin for nunit which catches log messages from log4net/nlog, and each message is assigned with specific test.
Currently to do it I use `TestContext.Progress.WriteLine()` and catch a messages at `public void OnTestEvent(string report)` from `ITestEventListener`. It works, but all messages are coming to the xml report and console output. Please advice how I can reach "silent" mode for my use case. Thx.

**Comment by jnm2** [comment]
@nvborisenko That reminds me of the breadcrumbs feature that @AGBrown suggested now that properties are read-only. I've also wanted a way to leave information or messages behind in the test results for diagnostic purposes without the console runner displaying it except on failure.

@CharliePoole Can you help me wrap my head around this? Why should Progress write to stderr by default? If the console runner intercepts stderr, how does the console runner differentiate between progress messages and error messages?

**Comment by CharliePoole** [comment]
Do I understand correctly that you want a special communication channel from a test to an extension? The existing facility was created for console output, of course, so that's what it is tailored for.

The current issue asks to expand the facility but retains the notion that it is all immediate console output. At least I think that's the intention - I'm waiting for @OsirisTerje to clarify it.

In any case, it seems to me that you want a special channel for handling logging. You might be better served to start a new issue.

I also suggest you examine how NUnit V2 handled log4net capture. There was no code needed in the framework to do it. I'll find and post some links when I'm at my computer.

**Comment by nvborisenko** [comment]
Right, this is kind of special channel between test and extension what is hidden from user. Imagine we have 2 superlong tests which are executing in parallel. Both of them write something into logger, for instance using log4net. We have an extension for nunit which handles all log messages, and it works good. The extension knows relation between message and test. Big problem here is that all log messages come to console, and any CI tool is overloaded. It's not big issue if tests produce textual logging, but it's big issue if tests produce some another kind of information like screenshots/videos/etc. Hidden channel will help to resolve this issue, nunit can be effectively extended by advanced custom reporters.

**Comment by CharliePoole** [comment]
@OsirisTerje I'm still confused about this request. Are you wanting NUnit to have it's own internal logging facility, such that the user sets a level and only that level of messages should be displayed? That would make some sense, but piggybacking it on top of stdOut or stdErr output doesn't seem to make a lot of sense.

Of course, with the four channels, the runner could maintain the desired level of output, but that means every log statement would need to be executed, including any code involved in formatting the message, because the framework would not know the desired level. IMO, if that's the kind of facility you want, it could easily be built inside the framework, separate from the kinds of console output that are now defined as __always__ appearing. Excuse me if I misunderstood your intent.

**Comment by CharliePoole** [comment]
@nvborisenko Currently, we define three sorts of output test text output:

1. Standard output, which goes into the test result for the test that sends it and is therefore delayed until the end of the test.
2. Immediate output, which goes directly to the runner, to be do be displayed in any way the runner chooses.
   2.1 Error output
   2.2 Non-error output (called "Progress")

Type 2 (immediate) is built on top of what is basically a multi-channel communications facility. IMO, we shouldn't do anything to confuse the user-facing output with the underlying facility. (The confusion I'm thinking about is primarily that of less-experienced users.)

I don't think that something to be used exclusively by a particular extension should be part of the public-facing API that you develop for logging. It would then be a temptation for users who see it in the intellisense to try it out, but of course nothing would happen without your extension.

OTOH, it would make perfect sense to have such a facility embedded within the Internal methods that NUnit exposes for your use - that is, public Types in the Internal namespace. Your extensions can naturally access those types and methods but users do not see them without deliberately adding a using statement.

In the case of text output, however, I'm afraid we didn't do a very good job of separating the public-facing part from the internal implementation. For example, `TestContext.Progress` points directly to a TextWriter that goes to stdErr, relying on the internal mechanisms that capture standard error and redirect it to a communication facility. This should probably be refactored.

Meanwhile, you might consider creating your own `EventListenerTextWriter` with a special-purpose steam name and using stdErr as the destination. The result of this may vary for each runner, depending on how the runner handles unknown stream names. For example, if memory serves, I think the console runner ignores the stream name except for coloring "Error" red. Other runners may only display "Progress" and "Error".

Feel free to contact me directly if you need help or input about your extension. I'm no longer working actively here, but I'm glad to share what I know.

**Comment by nvborisenko** [comment]
@CharliePoole thanks for your explanations. Could you please help with real example?
https://github.com/reportportal/example-net-nunit
This is simple tests project with nunit tests. Also there is extension which implements `ITestEventListener` interface. The responsibility of the extension is to listen execution events and send data to custom report server in real-time. Tests are using log4net to log messages (see Class1 -> Test1). Log4Net (actually appender) sends all messages to the nunit extension via `TestContext.Progress.WriteLine()`, so extension is able associate the message with test.
If you run tests in this example via `build.cmd` (build solution, configure nunit.console and run tests), you can see that all log messages also appear in console. In most cases this is OK, because log messages belongs to user and user wants to see them as a part of test. But in my example log4net appender talks to nunit extension: "Hey all nunit extensions, I have a message, take it, you know what to do with it". In this case this kind of communication is considered as internal or diagnostic communication between test domain and nunit app domain. It should be hidden from user. If user wants to see log messages in console, they just add `ConsoleAppender` and are happy.
Sure, to communicate between processes we can implement own mechanism, but first of all we want to reuse nunit stuff which partly fits our requirements.
For comparison [xunit](https://xunit.github.io/docs/capturing-output) has separate class for hidden messages - `DiagnosticMessage`, and user is able to configure runner whether to show or not these type of messages.
@CharliePoole could you please share your experience and suggest the best way how to resolve our issue? Thanks in advance.

**Comment by CharliePoole** [comment]
I'm not suggesting you should develop your own mechanism. What I'm saying is that the NUnit mechanism you have chosen serves a very specific purpose: to display progress messages so that the user knows his test has not died. It was requested, and we added it for just that purpose.

So I don't think we change how that works. A progress message that is hidden makes little sense.

And it's trivial for you to use the same underlying implementation that is used by the progress method to achieve your goal. All you have to do is create an `EventListenerTextWriter` in your extension, with a stream name known to you alone and (for the time being) a base writer of `Console.Error`. This will be distributed by NUnit using your own stream name. Only you will react to that name. However, it will probably cause some bug reports because certain runners are most likely not checking the stream name.

A more long term solution would be to define an XML message element different from "textoutput" for specific use in messaging between components.

I do feel that we are going well out of scope for this issue. Can we discuss either offline or on an issue that's for your exact problem?

**Comment by nvborisenko** [comment]
Created separate issue: https://github.com/nunit/nunit-console/issues/433

**Comment by AGBrown** [comment]
> That reminds me of the breadcrumbs feature that @AGBrown suggested now that properties are read-only. I've also wanted a way to leave information or messages behind in the test results for diagnostic purposes without the console runner displaying it except on failure.

I've opened an issue to describe this: #2915
